### PR TITLE
API cleanup - adding last_metering

### PR DESCRIPTION
**Pull request by lechup**
* added CACHES to settings
* added Station.last_metering with cache
* added first tests for StationAPI
* fix of FuzzyFloat in factories, created FuzzyFloatRound
* some other minior factories fixes

---
This change is [Reviewable](https://reviewable.io/reviews/enviromonitor/enviromonitorweb/49)

### CODE REVIEW

**Comment by khasinski** [comment]
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[api/serializers.py, line 79 at r1](https://reviewable.io:443/reviews/enviromonitor/enviromonitorweb/49#-KcSVeu13zEBGFbyrxXu:-KcSVeu13zEBGFbyrxXv:b-xmptxp) ([raw file](https://github.com/enviromonitor/enviromonitorweb/blob/ce8c455b34ebff58559a9aa4f2b5717c7fbb0137/api/serializers.py#L79)):*
> ```Python
>             'city',
>             'district',
>             'owner'
> ```

Append `,` at the end of list

---

*[api/tests/factories.py, line 28 at r1](https://reviewable.io:443/reviews/enviromonitor/enviromonitorweb/49#-KcSVkaide52DCX2lqry:-KcSVkaide52DCX2lqrz:b-i9qqj8) ([raw file](https://github.com/enviromonitor/enviromonitorweb/blob/ce8c455b34ebff58559a9aa4f2b5717c7fbb0137/api/tests/factories.py#L28)):*
> ```Python
>         if self.ndigits:
>             return round(fuzz, self.ndigits)
>         else:
> ```

`else` is not needed here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/enviromonitor/enviromonitorweb/49)*

**Comment by grillazz** [comment]
@lechup what is the purpose of adding cache with `def last_metering` ? I afraid I don't understand this idea ? Why not implement Redis as proper cache ?

**Comment by lechup** [comment]
@grillazz

You can always change caching backend later by changing caches settings key.

I wanted to cache `last_metering` to avoid many db queries on station listing. I've asked on slack if You have any better idea. Now each row in results it's a additional db query and it is quite big bottleneck.

This `last_metering` cache could be invalidated on new Metering creation for each Station and be cached forever when results are fetched from API.

I see two more options to consider:
* use django-denorm + bg jobs
* use django-cachalot

Hope it is clear now.

**Comment by lechup** [comment]
Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[api/serializers.py, line 79 at r1](https://reviewable.io:443/reviews/enviromonitor/enviromonitorweb/49#-KcSVeu13zEBGFbyrxXu:-KdQmmf0rNdLbwj-LsNP:b-896fix) ([raw file](https://github.com/enviromonitor/enviromonitorweb/blob/ce8c455b34ebff58559a9aa4f2b5717c7fbb0137/api/serializers.py#L79)):*
<details><summary><i>Previously, khasinski (Krzysztof Hasiński) wrote…</i></summary><blockquote>

Append `,` at the end of list
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/enviromonitor/enviromonitorweb/49)*

File: api/tests/test_views.py [hunk 1]
```diff
@@ -89,3 +100,86 @@ def test_project_detail_delete(self):
...
+        station_api_url = reverse('station-detail', kwargs={'pk': station.pk})
+
+        metering = MeteringFactory.build(station=None)
+        metering_data = MeteringSerializer(metering).data
+
+        # test cache key removal before add_metering
+        self.assertEqual(station.last_metering, MeteringSerializer(None).data)
+
+        self.assertEqual(station.metering_set.count(), 0)
+        add_metering_api_url = '{}add_metering/?token={}'.format(station_api_url, station.token)
```
Comment (khasinski) [inline api/tests/test_views.py]: URLs should use `-` for word separation

File: api/views.py [hunk 2]
```diff
@@ -15,6 +20,28 @@ class StationViewSet(ModelViewSet):
     serializer_class = StationSerializer
     filter_class = StationFilterSet
 
+    @detail_route(methods=['post'], permission_classes=[AllowAny])
+    def add_metering(self, request, pk=None):
+        station = self.get_object()
+        request_token = request.query_params.get('token')
+        if request_token is None or station.token != request_token:
+            raise StationWrongToken
+        else:
```
Comment (khasinski) [inline api/views.py]: `else`  is not needed here

File: api/tests/test_views.py (hunk 1 above)
Comment (lechup) [inline api/tests/test_views.py]: :+1: it's hidden inside drf documentation how to do it properly, but manage to find it: http://www.django-rest-framework.org/api-guide/routers/#extra-link-and-actions

File: api/views.py (hunk 2 above)
Comment (lechup) [inline api/views.py]: :+1:

**Comment by grillazz** [comment]
Reviewed 1 of 6 files at r1, 6 of 8 files at r3, 1 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/enviromonitor/enviromonitorweb/49)*

**Comment by lechup** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[api/tests/factories.py, line 28 at r1](https://reviewable.io:443/reviews/enviromonitor/enviromonitorweb/49#-KcSVkaide52DCX2lqry:-KdVlZ1PCzRsJMT-xZYW:b-896fix) ([raw file](https://github.com/enviromonitor/enviromonitorweb/blob/ce8c455b34ebff58559a9aa4f2b5717c7fbb0137/api/tests/factories.py#L28)):*
<details><summary><i>Previously, khasinski (Krzysztof Hasiński) wrote…</i></summary><blockquote>

`else` is not needed here
</blockquote></details>

Done.

---

*[EnviroMonitorWeb/settings/base.py, line 135 at r1](https://reviewable.io:443/reviews/enviromonitor/enviromonitorweb/49#-KcSSZbtSXjgsGBoXXTM-r1-135:-KdVkTEClX4-HqQiWrp3:be5th3b) ([raw file](https://github.com/enviromonitor/enviromonitorweb/blob/ce8c455b34ebff58559a9aa4f2b5717c7fbb0137/EnviroMonitorWeb/settings/base.py#L135)):*
<details><summary><i>Previously, grillazz (Jakub Miazek) wrote…</i></summary><blockquote>

Did you considered Redis as cache backend ?
</blockquote></details>

I would pick memcached - we do not need to keep data in cache in case of power outage, and it is designed for cacheing - but still this is rather env setting to me, it can be introduced later in proper env file, as a default local-mem-cache is quite reasonable for me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/enviromonitor/enviromonitorweb/49)*

**Comment by grillazz** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/enviromonitor/enviromonitorweb/49#-:-KdVnJpwGnsKoziSKs5E:bnfp4nl)*

**Pull request merged**